agent/envoyextensions/builtin/otel-access-logging/structs.go

```diff
@@ -0,0 +1,434 @@
+	cmn "github.com/hashicorp/consul/envoyextensions/extensioncommon"
+	"github.com/hashicorp/go-multierror"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+const (
+	LocalAccessLogClusterName = "local_access_log"
+
+	localhost     = "localhost"
+	localhostIPv4 = "127.0.0.1"
+	localhostIPv6 = "::1"
+)
+
+type AccessLog struct {
+	LogName                 string
+	GrpcService             *GrpcService
+	BufferFlushInterval     *time.Duration
+	BufferSizeBytes         uint32
+	FilterStateObjectsToLog []string
+	RetryPolicy             *RetryPolicy
+	Body                    interface{}
+	Attributes              map[string]interface{}
+	ResourceAttributes      map[string]interface{}
+}
+
+func (a *AccessLog) normalize() error {
+	if a.GrpcService != nil {
+		a.GrpcService.normalize()
+	} else {
+		return fmt.Errorf("missing GrpcService")
+	}
+
+	if a.RetryPolicy != nil {
+		a.RetryPolicy.normalize()
+	}
+
+	return nil
+}
+
+func (a *AccessLog) validate() error {
+	if a == nil {
+		return nil
+	}
+
+	a.normalize()
+
+	var resultErr error
+
+	var field string
+	var validate func() error
+	field = "GrpcService"
+	validate = a.GrpcService.validate
+
+	if err := validate(); err != nil {
+		resultErr = multierror.Append(resultErr, fmt.Errorf("failed to validate Config.%s: %w", field, err))
+	}
+
+	return resultErr
+}
+
+func (a *AccessLog) envoyGrpcService(cfg *cmn.RuntimeConfig) (*envoy_core_v3.GrpcService, error) {
+	target := a.GrpcService.Target
+	clusterName, err := a.getClusterName(cfg, target)
+	if err != nil {
+		return nil, err
+	}
+
+	var initialMetadata []*envoy_core_v3.HeaderValue
+	for _, meta := range a.GrpcService.InitialMetadata {
+		initialMetadata = append(initialMetadata, meta.toEnvoy())
+	}
+
+	return &envoy_core_v3.GrpcService{
+		TargetSpecifier: &envoy_core_v3.GrpcService_EnvoyGrpc_{
+			EnvoyGrpc: &envoy_core_v3.GrpcService_EnvoyGrpc{
+				ClusterName: clusterName,
+				Authority:   a.GrpcService.Authority,
+			},
+		},
+		Timeout:         target.timeoutDurationPB(),
+		InitialMetadata: initialMetadata,
+	}, nil
+}
+
+// getClusterName returns the name of the cluster for the OpenTelemetry access logging service.
+// If the extension is configured with an upstream OpenTelemetry access logging service then the name of the cluster for
+// that upstream is returned. If the extension is configured with a URI, the only allowed host is `localhost`
+// and the extension will insert a new cluster with the name "local_access_log", so we use that name.
+func (a *AccessLog) getClusterName(cfg *cmn.RuntimeConfig, target *Target) (string, error) {
+	var err error
+	clusterName := LocalAccessLogClusterName
+	if target.isService() {
+		if clusterName, err = target.clusterName(cfg); err != nil {
+			return "", err
+		}
+	}
+	return clusterName, nil
+}
+
+func (a *AccessLog) isGRPC() bool {
+	return a.GrpcService != nil
+}
```

Comment:
OTEL only works over gRPC so I think we can remove this func.

Reply:
Removed the method